prov/util/src/util_cq.c

```diff
@@ -570,68 +549,265 @@ void ofi_cq_progress(struct util_cq *cq)
+	}
+
+	cq->peer_cq->fid.fclass = FI_CLASS_PEER_CQ;
+	cq->peer_cq->fid.context = cq;
+	cq->peer_cq->fid.ops = &util_peer_cq_fi_ops;
+
+	return FI_SUCCESS;
+free:
+	free(cq->peer_cq);
+	return ret;
+}
+
+int ofi_cq_init(const struct fi_provider *prov, struct fid_domain *domain,
+		 struct fi_cq_attr *attr, struct util_cq *cq,
+		 ofi_cq_progress_func progress, void *context)
+{
+	struct fi_wait_attr wait_attr;
+	struct fid_wait *wait;
+	enum ofi_lock_type lock_type;
+	int ret;
+
+	assert(progress);
+	ret = ofi_check_cq_attr(prov, attr);
 	if (ret)
 		return ret;
 
-	/* CQ must be fully operational before adding to wait set */
-	if (cq->wait) {
-		ret = fi_poll_add(&cq->wait->pollset->poll_fid,
-				  &cq->cq_fid.fid, 0);
+	cq->cq_fid.fid.ops = &util_cq_fi_ops;
+	cq->cq_fid.ops = &util_cq_ops;
+	cq->progress = progress;
+
+	cq->domain = container_of(domain, struct util_domain, domain_fid);
+	ofi_atomic_initialize32(&cq->ref, 0);
+	ofi_atomic_initialize32(&cq->wakeup, 0);
+	dlist_init(&cq->ep_list);
+	ret = ofi_mutex_init(&cq->ep_list_lock);
+	if (ret)
+		return ret;
+
+	if (cq->domain->threading == FI_THREAD_COMPLETION ||
+	    cq->domain->threading == FI_THREAD_DOMAIN)
+		lock_type = OFI_LOCK_NOOP;
+	else
+		lock_type = cq->domain->lock.lock_type;
+
+	ret = ofi_genlock_init(&cq->cq_lock, lock_type);
+	if (ret)
+		goto destroy1;
+
+	cq->flags = attr->flags;
+	cq->cq_fid.fid.fclass = FI_CLASS_CQ;
+	cq->cq_fid.fid.context = context;
+
+	if (attr->flags & FI_PEER) {
+		cq->peer_cq = ((struct fi_peer_cq_context *) context)->cq;
+		cq->cq_fid.ops = &util_peer_cq_ops;
+	} else {
+		ret = util_init_peer_cq(cq, attr);
 		if (ret)
-			goto cleanup;
+			goto destroy2;
 	}
 
-	cq->cirq = util_comp_cirq_create(attr->size == 0 ? UTIL_DEF_CQ_SIZE : attr->size);
-	if (!cq->cirq) {
-		ret = -FI_ENOMEM;
+	ofi_atomic_inc32(&cq->domain->ref);
+	switch (attr->wait_obj) {
+	case FI_WAIT_NONE:
+		wait = NULL;
+		break;
+	case FI_WAIT_UNSPEC:
+	case FI_WAIT_FD:
+	case FI_WAIT_POLLFD:
+	case FI_WAIT_MUTEX_COND:
+	case FI_WAIT_YIELD:
+		memset(&wait_attr, 0, sizeof wait_attr);
+		wait_attr.wait_obj = attr->wait_obj;
+		cq->internal_wait = 1;
+		ret = fi_wait_open(&cq->domain->fabric->fabric_fid,
+				   &wait_attr, &wait);
+		if (ret)
+			goto cleanup;
+		break;
+	case FI_WAIT_SET:
+		wait = attr->wait_set;
+		break;
+	default:
+		assert(0);
 		goto cleanup;
 	}
 
-	if (cq->domain->info_domain_caps & FI_SOURCE) {
-		cq->src = calloc(cq->cirq->size, sizeof *cq->src);
-		if (!cq->src) {
-			ret = -FI_ENOMEM;
+	/* CQ must be fully operational before adding to wait set */
+	if (wait) {
+		cq->wait = container_of(wait, struct util_wait, wait_fid);
+		ret = fi_poll_add(&cq->wait->pollset->poll_fid,
+				  &cq->cq_fid.fid, 0);
+		if (ret) {
+			if (cq->internal_wait) {
+				fi_close(&cq->wait->wait_fid.fid);
+				cq->wait = NULL;
+			}
 			goto cleanup;
 		}
 	}
 	return 0;
-
+destroy2:
+	ofi_genlock_destroy(&cq->cq_lock);
+destroy1:
+	ofi_mutex_destroy(&cq->ep_list_lock);
```

Comment:
The cleanup order with destroy1 and destroy2 look correct.  But the jump to cleanup when an error occurs even further into the function (that is, we skip the destroy cleanups), looks wrong.  I guess I need to go see what ofi_cq_cleanup() actually does relative to the above changes.

Reply:
I think the skipping the destroys is fine but I think the fallthroughs from the destroys are not right. Gah I'll tweak it again. Sorry

Reply:
Updated again to fix cleanup path. I think it should be good this time 🤞